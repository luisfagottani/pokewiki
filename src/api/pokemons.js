import { fetchApi } from './api';

const Pokemons = {
  getAllPokemons: async () => {
    return fetchApi('pokemon', {
      method: 'GET',
    });
  },
};

export default Pokemons;
