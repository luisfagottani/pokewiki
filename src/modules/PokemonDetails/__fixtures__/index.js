const details = {
  name: 'bulbasaur',
  abilities: [
    {
      ability: {
        name: 'overgrow',
        url: 'https://pokeapi.co/api/v2/ability/65/',
      },
      is_hidden: false,
      slot: 1,
    },
    {
      ability: {
        name: 'chlorophyll',
        url: 'https://pokeapi.co/api/v2/ability/34/',
      },
      is_hidden: true,
      slot: 3,
    },
  ],
  height: 7,
  stats: [
    {
      base_stat: 45,
      effort: 0,
      stat: {
        name: 'hp',
        url: 'https://pokeapi.co/api/v2/stat/1/',
      },
    },
    {
      base_stat: 49,
      effort: 0,
      stat: {
        name: 'attack',
        url: 'https://pokeapi.co/api/v2/stat/2/',
      },
    },
    {
      base_stat: 49,
      effort: 0,
      stat: {
        name: 'defense',
        url: 'https://pokeapi.co/api/v2/stat/3/',
      },
    },
    {
      base_stat: 65,
      effort: 1,
      stat: {
        name: 'special-attack',
        url: 'https://pokeapi.co/api/v2/stat/4/',
      },
    },
    {
      base_stat: 65,
      effort: 0,
      stat: {
        name: 'special-defense',
        url: 'https://pokeapi.co/api/v2/stat/5/',
      },
    },
    {
      base_stat: 45,
      effort: 0,
      stat: {
        name: 'speed',
        url: 'https://pokeapi.co/api/v2/stat/6/',
      },
    },
  ],
  weight: 69,
  id: 1,
  base_experience: 64,
};

export { details };
