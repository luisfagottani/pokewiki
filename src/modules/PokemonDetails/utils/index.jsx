const progressBarColor = (progress) => {
  if (progress < 60) {
    return 'blue';
  } else if (progress >= 60 && progress <= 70) {
    return 'yellow';
  } else {
    return 'red';
  }
};

export { progressBarColor };
