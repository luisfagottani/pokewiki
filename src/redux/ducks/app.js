import produce from 'immer';

export const REDUCER_NAME = 'app';

// TYPES
export const Types = {
  SET_GLOBAL_LOADING: `${REDUCER_NAME}/SET_GLOBAL_LOADING`,
};

// INITIAL STATE
const initialState = {
  isLoading: true,
  locale: 'pt-BR',
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.SET_GLOBAL_LOADING:
      draft.isLoading = action.payload;
      return draft;

    case Types.SET_USER_PROFILE:
      draft.profile = action.profile;
      return draft;

    default:
      return draft;
  }
});

// ACTION FUNCTIONS

export function defineGlobalLoading(status) {
  return {
    type: Types.SET_GLOBAL_LOADING,
    payload: status,
  };
}

// SELECTOR FUNCTIONS
export function getGlobalLoading(state) {
  return state[REDUCER_NAME].isLoading;
}

export function getLocale(state) {
  return state[REDUCER_NAME].locale;
}
