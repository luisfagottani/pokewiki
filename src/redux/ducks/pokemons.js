import produce from 'immer';

export const REDUCER_NAME = 'pokemons';

// TYPES
export const Types = {
  FETCH_POKEMONS: `${REDUCER_NAME}/FETCH_POKEMONS`,
  PERSIST_POKEMONS: `${REDUCER_NAME}/PERSIST_POKEMONS`,
};

// INITIAL STATE
const initialState = {
  species: [],
};

// REDUCER
export const reducer = produce((draft = initialState, action) => {
  switch (action.type) {
    case Types.PERSIST_POKEMONS:
      draft.species.push(...action.payload);

      return draft;
    default:
      return draft;
  }
});

// ACTION FUNCTIONS
export function fetchPokemons() {
  return {
    type: Types.FETCH_POKEMONS,
  };
}

export function persistPokemons(pokemons) {
  return {
    type: Types.PERSIST_POKEMONS,
    payload: pokemons,
  };
}

export function getPokemons(state) {
  return state[REDUCER_NAME].species;
}
