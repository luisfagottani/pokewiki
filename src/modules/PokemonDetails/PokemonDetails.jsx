import { connect } from 'react-redux';

import Layout from 'modules/shared/Layout/Layout';
import { fetchPokemonById, getPokemonById } from 'redux/ducks/pokemons';
import Profile from './components/Profile';
import React from 'react';

const PokemonDetails = ({ pokemon, handleFetchPokemonById, pokeId }) => {
  React.useEffect(() => {
    if (!pokemon || !pokemon?.isFetched) {
      handleFetchPokemonById(pokeId);
    }
  }, [pokemon, pokeId, handleFetchPokemonById]);

  return (
    <Layout.Row>
      <Layout.Column type="one-quarter">
        <Profile name={pokemon?.name} />
      </Layout.Column>
      <Layout.Column type="auto">asdadsa</Layout.Column>
    </Layout.Row>
  );
};

const mapStateToProps = (state, props) => ({
  pokemon: getPokemonById(state, props.pokeId),
});

const mapDispatchToProps = {
  handleFetchPokemonById: fetchPokemonById,
};

export default connect(mapStateToProps, mapDispatchToProps)(PokemonDetails);
