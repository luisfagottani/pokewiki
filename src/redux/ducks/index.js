import { combineReducers } from 'redux';

import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from './app';
import { reducer as PokemonsReducer, REDUCER_NAME as APP_POKEMONS_REDUCER } from './pokemons';
import { reducer as GenerationsReducer, REDUCER_NAME as APP_GENERATIONS_REDUCER } from './generations';

export default combineReducers({
  [APP_REDUCER_NAME]: AppReducer,
  [APP_POKEMONS_REDUCER]: PokemonsReducer,
  [APP_GENERATIONS_REDUCER]: GenerationsReducer,
});
