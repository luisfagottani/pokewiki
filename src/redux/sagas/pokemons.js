/* eslint-disable import/no-anonymous-default-export */
import { call, takeLatest, put, delay } from '@redux-saga/core/effects';
import PokemonsApi from 'api/pokemons';
import { defineGlobalLoading } from 'redux/ducks/app';
import { getPokemons, persistPokemons, Types } from 'redux/ducks/pokemons';

export function* persistInStore(species) {
  yield put(persistPokemons(species));
  yield put(defineGlobalLoading(false));
}

// export function* updatePokemons(pokemon) {
//   const state = yield select();
//   let pokemons = { ...getPokemons(state) };
//   const { name, ...details } = pokemon;

//   if (pokemons[name]) {
//     pokemons[name] = { details, isFetched: true, ...pokemons[name] };
//   } else {
//     pokemons[name] = { name, details, isFetched: true };
//   }

//   return pokemons;
// }

export function* fetchPokemons() {
  try {
    yield put(defineGlobalLoading(true));
    yield delay(1000);
    const response = yield call([PokemonsApi, PokemonsApi.getAllPokemons], 0, 200);
    const species = response?.results.reduce((current, next) => {
      const name = next?.name;
      if (!current[name]) {
        current[name] = { name, details: {}, isFetched: false };
      }
      return current;
    }, {});

    yield persistInStore(species);
  } catch (error) {
    console.log(error);
  }
}

export function* fetchPokemonById(id) {
  try {
    const response = yield call([PokemonsApi, PokemonsApi.getPokemonById], id);
    // const species = yield updatePokemons(response);
    // yield persistInStore(species);
  } catch (error) {
    console.log(error);
  }
}

export default function* () {
  yield takeLatest(Types.FETCH_POKEMONS, function* (data) {
    yield call(fetchPokemons, data.payload);
  });

  yield takeLatest(Types.FETCH_POKEMON_BY_ID, function* (data) {
    yield call(fetchPokemonById, data.payload);
  });
}
