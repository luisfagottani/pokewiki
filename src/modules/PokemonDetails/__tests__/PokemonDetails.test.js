import { createMockStore, render } from 'utils/tests';
import { reducer as AppReducer, REDUCER_NAME as APP_REDUCER_NAME } from 'redux/ducks/app';
import { reducer as PokemonsReducer, REDUCER_NAME as POKEMON_REDUCER_NAME } from 'redux/ducks/pokemons';
import PokemonDetails from '../PokemonDetails';
import { details } from '../__fixtures__';

const mockStore = createMockStore({
  reducers: { [APP_REDUCER_NAME]: AppReducer, [POKEMON_REDUCER_NAME]: PokemonsReducer },
  initialState: {
    [APP_REDUCER_NAME]: {
      isLoading: true,
      locale: 'EN',
    },
    [POKEMON_REDUCER_NAME]: {
      species: { bulbasaur: { name: 'bulbasaur', details: details } },
    },
  },
});

describe('<PokemonDetails />', () => {
  it('Should render without errors', () => {
    const { container } = render(<PokemonDetails />, mockStore);

    expect(container).toMatchSnapshot();
  });

  it('Should render all sections', () => {
    const { getByTestId } = render(<PokemonDetails />, mockStore);

    getByTestId('profile');
    getByTestId('abilities');
    getByTestId('stats');
  });
});
