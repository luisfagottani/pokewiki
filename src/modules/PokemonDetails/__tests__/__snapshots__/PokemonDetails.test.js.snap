// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonDetails /> Should render without errors 1`] = `
<div>
  <section
    class="layout__row"
  >
    <div
      class="layout__column layout__column--one-quarter"
    >
      <div
        class="layout__content"
      >
        <div
          class="profile"
          data-testid="profile"
        >
          <figure
            class="profile__figure"
          >
            <figcaption
              class="profile__caption"
            />
            <svg
              class="profile__image"
            >
              pokeball-icon.svg
            </svg>
          </figure>
          <div
            class="list list--loading"
          >
            <dl
              class="list__dl"
            >
              <dt
                class="list__dt"
              >
                National N¬∫
              </dt>
              <dd
                class="list__dd"
              />
              <dt
                class="list__dt"
              >
                Base Experience
              </dt>
              <dd
                class="list__dd"
              />
              <dt
                class="list__dt"
              >
                Height
              </dt>
              <dd
                class="list__dd"
              />
              <dt
                class="list__dt"
              >
                Weight
              </dt>
              <dd
                class="list__dd"
              />
            </dl>
          </div>
        </div>
      </div>
    </div>
    <div
      class="layout__column layout__column--auto"
    >
      <div
        class="layout__content"
      >
        <div
          class="abilities"
          data-testid="abilities"
        >
          <h2
            class="layout__subtitle"
          >
            üïπÔ∏è Abilities
          </h2>
          <div
            class="loading"
            data-testid="loading-section"
          >
            <svg
              class="loading__ball"
            >
              pokeball-icon.svg
            </svg>
          </div>
          <ul
            class="abilities__list"
          />
        </div>
        <div
          class="stats"
          data-testid="stats"
        >
          <h2
            class="layout__subtitle"
          >
            üìà Base Stats
          </h2>
          <div
            class="loading"
            data-testid="loading-section"
          >
            <svg
              class="loading__ball"
            >
              pokeball-icon.svg
            </svg>
          </div>
          <ul
            class="stats__list"
          />
        </div>
      </div>
    </div>
  </section>
</div>
`;
